File: internal/login/radar.go
package login

import (
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/mxschmitt/playwright-go"

	"opencode-go-manager/internal/cline"
	"opencode-go-manager/internal/config"
	"opencode-go-manager/internal/herosms"
	"opencode-go-manager/internal/model"
	"opencode-go-manager/internal/usage"
)

const radarPhoneAttempts = 3

func handleRadar(page playwright.Page, cfg config.Config, log Logger) error {
	if !strings.Contains(googlePath(page.URL()), "radar-challenge") || onIdentityProvider(page.URL()) {
		return nil
	}
	if err := waitRadarForm(page); err != nil {
		return err
	}
	if cfg.HeroSMSAPIKey == "" {
		return fmt.Errorf("遇到手机验证，请先在设置里填写 Hero SMS API Key")
	}
	if cfg.HeroSMSCountry <= 0 {
		return fmt.Errorf("遇到手机验证，请先在设置里选择接码区域和报价")
	}

	sendURL := page.URL()
	if strings.Contains(sendURL, "radar-challenge/verify") {
		sendURL = radarSendURL(sendURL)
	}
	if err := gotoRadarSend(page, sendURL, log); err != nil {
		return err
	}
	sendURL = page.URL()

	sms := herosms.New(cfg.HeroSMSAPIKey, cfg.HeroSMSService)
	for i := 1; i <= radarPhoneAttempts; i++ {
		err := requestRadarCode(page, cfg, sms, log, i)
		if err == nil {
			return nil
		}
		if !shouldRetryRadarPhone(err) {
			return err
		}
		log("第 %d/%d 次接码失败: %v", i, radarPhoneAttempts, err)
		if i < radarPhoneAttempts {
			log("返回手机号输入页，换一个 Hero SMS 号码重试")
			sleep(2000)
			if err := gotoRadarSend(page, sendURL, log); err != nil {
				return err
			}
		}
	}
	return ErrSMSNeedRelogin
}

func requestRadarCode(page playwright.Page, cfg config.Config, sms *herosms.Client, log Logger, attempt int) error {
	if err := waitRadarForm(page); err != nil {
		return err
	}
	log("Hero SMS 取号 country=%d price=%s（第 %d/%d 次）", cfg.HeroSMSCountry, formatPrice(cfg.HeroSMSMaxPrice), attempt, radarPhoneAttempts)
	num, err := sms.GetNumber(cfg.HeroSMSCountry, cfg.HeroSMSMaxPrice)
	if err != nil {
		return err
	}
	finished := false
	defer func() {
		if !finished {
			sms.Cancel(num.ID)
		}
	}()
	log("已取号 +%d %s id=%s", num.PhoneCode, num.LocalNumber, num.ID)
	sleep(200)

	if num.PhoneCode <= 0 || num.LocalNumber == "" {
		return fmt.Errorf("取到的号码无法拆分区号: %s", num.Phone)
	}
	cc := fmt.Sprintf("+%d", num.PhoneCode)
	if err := fillAuthkitField(page, `input[name="country_code"]`, cc); err != nil {
		return fmt.Errorf("填写区号失败: %w", err)
	}
	sleep(150)
	if err := fillAuthkitField(page, `input[name="local_number"]`, num.LocalNumber); err != nil {
		return fmt.Errorf("填写手机号失败: %w", err)
	}
	sleep(200)
	if err := clickOneOf(page, []string{
		`button[data-hak-cta][type="submit"]`,
		`button.ak-PrimaryButton[type="submit"]`,
		`button[type="submit"]`,
	}, 15000, log, "发送验证码"); err != nil {
		return err
	}
	if err := waitAnyURL(page, []string{"radar-challenge/verify"}, 30000); err != nil {
		if onRadarSend(page.URL()) || visible(page, `input[name="local_number"]`) {
			return fmt.Errorf("发送验证码失败，号码可能已被使用: %w", err)
		}
		return fmt.Errorf("没有进入验证码页: %w", err)
	}
	sleep(200)

	log("等待短信验证码")
	code, err := sms.WaitCode(num.ID, 2*time.Minute)
	if err != nil {
		return err
	}
	log("收到验证码")
	sleep(120)
	if err := fillOTP(page, code); err != nil {
		return err
	}
	deadline := time.Now().Add(45 * time.Second)
	for time.Now().Before(deadline) {
		if !strings.Contains(page.URL(), "radar-challenge") {
			sms.Finish(num.ID)
			finished = true
			return nil
		}
		sleep(400)
	}
	return fmt.Errorf("提交验证码后仍停在验证页，当前 URL=%s", page.URL())
}

func gotoRadarSend(page playwright.Page, sendURL string, log Logger) error {
	if onRadarSend(page.URL()) && visible(page, `input[name="local_number"]`) {
		return nil
	}
	target := strings.TrimSpace(sendURL)
	if target == "" || strings.Contains(target, "radar-challenge/verify") {
		target = radarSendURL(page.URL())
	}
	if log != nil && target != "" {
		log("打开手机号输入页")
	}
	if target != "" {
		if _, err := page.Goto(target, playwright.PageGotoOptions{
			WaitUntil: playwright.WaitUntilStateDomcontentloaded,
			Timeout:   playwright.Float(30000),
		}); err != nil {
			if log != nil {
				log("打开手机号页失败: %v，尝试后退", err)
			}
			_, _ = page.GoBack()
		}
	} else {
		_, _ = page.GoBack()
	}
	sleep(300)
	if onRadarSend(page.URL()) || visible(page, `input[name="local_number"]`) {
		return nil
	}
	return fmt.Errorf("无法回到手机号输入页，当前 URL=%s", page.URL())
}

func radarSendURL(raw string) string {
	return strings.Replace(raw, "radar-challenge/verify", "radar-challenge/send", 1)
}

func waitRadarForm(page playwright.Page) error {
	deadline := time.Now().Add(15 * time.Second)
	for time.Now().Before(deadline) {
		if visible(page, `input[name="country_code"]`) && visible(page, `input[name="local_number"]`) {
			return nil
		}
		sleep(200)
	}
	return fmt.Errorf("接码页没有可填的手机号输入框，当前 URL=%s", page.URL())
}

func onRadarSend(u string) bool {
	return strings.Contains(u, "radar-challenge/send")
}

func isNoSMS(err error) bool {
	if err == nil {
		return false
	}
	return errors.Is(err, herosms.ErrWaitCodeTimeout) ||
		errors.Is(err, herosms.ErrCancelled) ||
		errors.Is(err, herosms.ErrUnavailable)
}

func shouldRetryRadarPhone(err error) bool {
	if isNoSMS(err) {
		return true
	}
	if err == nil {
		return false
	}
	msg := err.Error()
	if strings.Contains(msg, "号码可能已被使用") {
		return true
	}
	return strings.Contains(msg, "等待 URL 超时") && strings.Contains(msg, "radar-challenge/send")
}

func fillOTP(page playwright.Page, code string) error {
	code = digitsOnly(code)
	if len(code) < 4 {
		return fmt.Errorf("验证码无效")
	}
	if len(code) > 6 {
		code = code[:6]
	}
	inputs := page.Locator(`input[data-test="otp-input"], input[data-index], .ak-Otp input`)
	n, err := inputs.Count()
	if err != nil || n == 0 {
		if err := fillAny(page, `input[name="code"], input[autocomplete="one-time-code"]`, code); err != nil {
			return fmt.Errorf("填写验证码失败")
		}
		return nil
	}
	for i := 0; i < n && i < len(code); i++ {
		loc := inputs.Nth(i)
		if err := humanType(page, loc, string(code[i])); err != nil {
			_ = loc.Press(string(code[i]), playwright.LocatorPressOptions{Timeout: playwright.Float(2000)})
		}
		sleep(20)
	}
	return nil
}

func handleTerms(page playwright.Page, log Logger) error {
	if !visible(page, `label#terms, #terms`) {
		return nil
	}
	log("勾选条款")
	if err := clickOneOf(page, []string{`label#terms`, `#terms`}, 8000, log, "勾选条款"); err != nil {
		return err
	}
	sleep(400)
	return clickOneOf(page, []string{
		`button[type="button"].w-full`,
		`button.bg-black[type="button"]`,
		`button[type="button"]`,
	}, 15000, log, "确认注册")
}

func captureClinePayment(page playwright.Page, log Logger) (string, error) {
	u := page.URL()
	if !strings.Contains(u, "/onboarding/") && !strings.Contains(u, "/checkout") {
		if _, err := page.Goto(cline.AppBase+"/onboarding/individual-plan", playwright.PageGotoOptions{
			WaitUntil: playwright.WaitUntilStateDomcontentloaded,
			Timeout:   playwright.Float(60000),
		}); err != nil {
			return "", fmt.Errorf("打开套餐页失败: %w", err)
		}
		sleep(250)
	}
	if cookieExpired(page.URL()) {
		return "", fmt.Errorf("Cookie 已失效，需要重新登录")
	}
	if !strings.Contains(page.URL(), "/checkout") {
		_ = clickOneOf(page, []string{
			`button:has(svg.lucide-chevron-right)`,
			`button.h-12`,
			`button[type="button"]:has(svg)`,
		}, 20000, log, "进入结账")
	}
	deadline := time.Now().Add(60 * time.Second)
	for time.Now().Before(deadline) {
		if src := iframePaymentSrc(page); src != "" {
			return src, nil
		}
		if isStripe(page.URL()) {
			return page.URL(), nil
		}
		sleep(500)
	}
	return "", fmt.Errorf("等待支付 iframe 超时")
}

func iframePaymentSrc(page playwright.Page) string {
	loc := page.Locator(`iframe[src*="stripe"], iframe[src*="checkout"], iframe`).First()
	ok, err := loc.IsVisible()
	if err != nil || !ok {
		return ""
	}
	src, err := loc.GetAttribute("src")
	if err != nil {
		return ""
	}
	src = strings.TrimSpace(src)
	if src == "" {
		return ""
	}
	if strings.Contains(src, "stripe") || strings.Contains(src, "checkout") || strings.HasPrefix(src, "https://") {
		return src
	}
	return ""
}

func createClineKey(cfg config.Config, cookie, existingKey, existingUser string, log Logger) (key, userID string, err error) {
	a := model.Account{CookieHeader: cookie, APIKey: existingKey, UserID: existingUser, WorkspaceID: existingUser}
	if err := usage.Hydrate(&a, cfg.Proxy); err != nil {
		if existingKey != "" {
			return existingKey, existingUser, nil
		}
		return "", "", err
	}
	if a.APIKey != "" && a.APIKey != existingKey {
		log("已创建 API Key %s...", maskKey(a.APIKey))
	}
	return a.APIKey, a.UserID, nil
}

func fillAuthkitField(page playwright.Page, selector, value string) error {
	if err := waitVisibleOpt(page, selector, 20000, false); err != nil {
		return err
	}
	loc := page.Locator(selector).First()
	waitOverlayGone(page)
	if err := humanType(page, loc, value); err != nil {
		return CompactError(err)
	}
	if authkitInputMatch(loc, value) {
		return nil
	}
	current, _ := loc.InputValue()
	typed := value
	if strings.HasPrefix(strings.TrimSpace(current), "+") && strings.HasPrefix(value, "+") {
		typed = strings.TrimPrefix(value, "+")
	}
	if err := humanType(page, loc, typed); err != nil {
		return CompactError(err)
	}
	if !authkitInputMatch(loc, value) {
		got, _ := loc.InputValue()
		return fmt.Errorf("页面是 %q，期望 %q", got, value)
	}
	return nil
}

func authkitInputMatch(loc playwright.Locator, want string) bool {
	got, err := loc.InputValue()
	if err != nil {
		return false
	}
	return digitsOnly(got) == digitsOnly(want)
}

func formatPrice(n float64) string {
	if n <= 0 {
		return "auto"
	}
	return fmt.Sprintf("%g", n)
}

func digitsOnly(s string) string {
	var b strings.Builder
	for _, r := range s {
		if r >= '0' && r <= '9' {
			b.WriteRune(r)
		}
	}
	return b.String()
}

File: worker/radar.py
import time

from errors import WorkerError
from herosms import Client
from pageutil import click_one_of, field_ready, input_value, sleep_ms, visible, wait_any_url
from protocol import log
from urls import on_radar_url, radar_send_url


def digits_only(s: str) -> str:
    return "".join(ch for ch in (s or "") if ch.isdigit())


def authkit_input_match(page, selector: str, want: str) -> bool:
    return digits_only(input_value(page, selector)) == digits_only(want)


def type_authkit_field(page, selector: str, value: str) -> None:
    loc = page.locator(selector).first
    try:
        loc.fill(value, timeout=12000)
        if authkit_input_match(page, selector, value):
            return
    except Exception:
        pass
    loc.click(timeout=8000)
    page.keyboard.press("Control+a")
    page.keyboard.press("Backspace")
    loc.type(value, delay=18, timeout=12000)


def fill_authkit_field(page, selector: str, value: str) -> None:
    from pageutil import wait_visible

    wait_visible(page, selector, 20000, stop_if_done=False)
    type_authkit_field(page, selector, value)
    if authkit_input_match(page, selector, value):
        return
    typed = value
    current = input_value(page, selector)
    if current.strip().startswith("+") and value.startswith("+"):
        typed = value[1:]
    type_authkit_field(page, selector, typed)
    if not authkit_input_match(page, selector, value):
        raise RuntimeError(f'页面是 "{input_value(page, selector)}"，期望 "{value}"')


def radar_form_ready(page) -> bool:
    return field_ready(page, 'input[name="country_code"]') and field_ready(page, 'input[name="local_number"]')


def wait_radar_form(page, timeout_ms: float = 15000) -> None:
    deadline = time.time() + timeout_ms / 1000.0
    while time.time() < deadline:
        if radar_form_ready(page):
            return
        sleep_ms(200)
    raise RuntimeError(f"接码页没有可填的手机号输入框，当前 URL={page.url}")


def fill_otp(page, code: str) -> None:
    code = digits_only(code)
    if len(code) < 4:
        raise RuntimeError("验证码无效")
    code = code[:6]
    log("填写验证码")
    first = 'input[data-test="otp-input"], input[data-index], .ak-Otp input, input[name="code"], input[autocomplete="one-time-code"]'
    page.click(first, timeout=12000)
    page.keyboard.press("Control+a")
    page.keyboard.press("Backspace")
    page.keyboard.type(code, delay=18)
    sleep_ms(120)
    try:
        click_one_of(
            page,
            ['button[data-hak-cta][type="submit"]', 'button.ak-PrimaryButton[type="submit"]', 'button[type="submit"]'],
            4000,
            "提交验证码",
        )
    except Exception:
        pass


def on_radar_send(u: str) -> bool:
    return "radar-challenge/send" in u


def goto_radar_send(page, send_url: str) -> None:
    if on_radar_send(page.url) and visible(page, 'input[name="local_number"]'):
        return
    target = (send_url or "").strip()
    if not target or "radar-challenge/verify" in target:
        target = radar_send_url(page.url)
    if target:
        log("打开手机号输入页")
        try:
            page.goto(target, wait_until="domcontentloaded", timeout=30000)
        except Exception as exc:
            log("打开手机号页失败: %s，尝试后退", exc)
            try:
                page.go_back()
            except Exception:
                pass
    else:
        try:
            page.go_back()
        except Exception:
            pass
    sleep_ms(300)
    if on_radar_send(page.url) or visible(page, 'input[name="local_number"]'):
        return
    raise RuntimeError(f"无法回到手机号输入页，当前 URL={page.url}")


def request_radar_code(page, settings: dict, sms: Client, attempt: int) -> None:
    wait_radar_form(page)
    country = int(settings.get("hero_sms_country") or 0)
    price = float(settings.get("hero_sms_max_price") or 0)
    log("Hero SMS 取号 country=%d price=%s（第 %d 次）", country, ("%g" % price) if price > 0 else "auto", attempt)
    num = sms.get_number(country, price)
    finished = False
    try:
        log("已取号 +%s %s id=%s", num["phone_code"], num["local_number"], num["id"])
        sleep_ms(200)
        if int(num.get("phone_code") or 0) <= 0 or not num.get("local_number"):
            raise RuntimeError(f'取到的号码无法拆分区号: {num.get("phone")}')
        cc = f'+{num["phone_code"]}'
        fill_authkit_field(page, 'input[name="country_code"]', cc)
        sleep_ms(150)
        fill_authkit_field(page, 'input[name="local_number"]', num["local_number"])
        sleep_ms(200)
        click_one_of(
            page,
            ['button[data-hak-cta][type="submit"]', 'button.ak-PrimaryButton[type="submit"]', 'button[type="submit"]'],
            15000,
            "发送验证码",
        )
        try:
            wait_any_url(page, ["radar-challenge/verify"], 30000)
        except Exception as exc:
            if on_radar_send(page.url) or visible(page, 'input[name="local_number"]'):
                raise RuntimeError("发送验证码失败，号码可能已被使用") from exc
            raise
        sleep_ms(200)
        log("等待短信验证码")
        code = sms.wait_code(num["id"], 120)
        log("收到验证码")
        sleep_ms(120)
        fill_otp(page, code)
        log("验证码已填，等待离开验证页")
        deadline = time.time() + 45
        last = 0.0
        while time.time() < deadline:
            u = page.url
            if "radar-challenge" not in u:
                sms.finish(num["id"])
                finished = True
                log("已离开接码页，当前 URL=%s", u)
                return
            if time.time() - last > 6:
                log("仍在验证页，当前 URL=%s", u)
                last = time.time()
            sleep_ms(200)
        raise RuntimeError(f"提交验证码后仍停在验证页，当前 URL={page.url}")
    finally:
        if not finished:
            sms.cancel(num["id"])


def is_no_sms(exc: Exception) -> bool:
    msg = str(exc)
    return isinstance(exc, TimeoutError) or "等待验证码超时" in msg or "接码已取消" in msg or "Hero SMS 暂时不可用" in msg


def should_retry_phone(exc: Exception) -> bool:
    if is_no_sms(exc):
        return True
    msg = str(exc)
    if "号码可能已被使用" in msg:
        return True
    return "等待 URL 超时" in msg and "radar-challenge/send" in msg


def handle_radar(page, settings: dict) -> None:
    if not on_radar_url(page.url):
        return
    if not (settings.get("hero_sms_api_key") or "").strip():
        raise RuntimeError("遇到手机验证，请先在设置里填写 Hero SMS API Key")
    if int(settings.get("hero_sms_country") or 0) <= 0:
        raise RuntimeError("遇到手机验证，请先在设置里选择接码区域和报价")
    wait_radar_form(page)
    send_url = page.url
    if "radar-challenge/verify" in send_url:
        send_url = radar_send_url(send_url)
    goto_radar_send(page, send_url)
    send_url = page.url
    sms = Client(settings.get("hero_sms_api_key") or "", settings.get("hero_sms_service") or "")
    last = None
    attempts = 3
    for i in range(1, attempts + 1):
        try:
            request_radar_code(page, settings, sms, i)
            return
        except Exception as exc:
            last = exc
            if not should_retry_phone(exc):
                raise
            log("第 %d/%d 次接码失败: %s", i, attempts, exc)
            if i < attempts:
                log("返回手机号输入页，换一个 Hero SMS 号码重试")
                sleep_ms(300)
                goto_radar_send(page, send_url)
    raise WorkerError("多次接码失败，需要重新登录", "sms_relogin") from last


def handle_terms(page) -> None:
    if not visible(page, "label#terms, #terms"):
        return
    log("勾选条款")
    click_one_of(page, ["label#terms", "#terms"], 8000, "勾选条款")
    sleep_ms(400)
    click_one_of(
        page,
        ['button[type="button"].w-full', "button.bg-black[type=button]", 'button[type="button"]'],
        15000,
        "确认注册",
    )
